aws/hll/aws-s3-transfer-manager/src/io/part_reader.rs

```diff
@@ -0,0 +1,337 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+use std::cmp;
+use std::ops::DerefMut;
+use std::sync::Mutex;
+
+use bytes::{Buf, Bytes, BytesMut};
+
+use crate::io::error::Error;
+use crate::io::path_body::PathBody;
+use crate::io::stream::RawInputStream;
+use crate::io::InputStream;
+use crate::MEBIBYTE;
+
+/// Builder for creating a `ReadPart` implementation.
+#[derive(Debug)]
+pub(crate) struct Builder {
+    stream: Option<RawInputStream>,
+    part_size: usize,
+}
+
+impl Builder {
+    pub(crate) fn new() -> Self {
+        Self {
+            stream: None,
+            part_size: 5 * MEBIBYTE as usize,
+        }
+    }
+
+    /// Set the input stream to read from.
+    pub(crate) fn stream(mut self, stream: InputStream) -> Self {
+        self.stream = Some(stream.inner);
+        self
+    }
+
+    /// Set the target part size that should be used when reading data.
+    ///
+    /// All parts except for possibly the last one should be of this size.
+    pub(crate) fn part_size(mut self, part_size: usize) -> Self {
+        self.part_size = part_size;
+        self
+    }
+
+    pub(crate) fn build(self) -> impl ReadPart {
+        let stream = self.stream.expect("input stream set");
+        match stream {
+            RawInputStream::Buf(buf) => {
+                PartReader::Bytes(BytesPartReader::new(buf, self.part_size))
+            }
+            RawInputStream::Fs(path_body) => {
+                PartReader::Fs(PathBodyPartReader::new(path_body, self.part_size))
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum PartReader {
+    Bytes(BytesPartReader),
+    Fs(PathBodyPartReader),
+}
+
+impl ReadPart for PartReader {
+    async fn next_part(&self) -> Result<Option<PartData>, Error> {
+        match self {
+            PartReader::Bytes(bytes) => bytes.next_part().await,
+            PartReader::Fs(path_body) => path_body.next_part().await,
+        }
+    }
+}
+
+/// Data for a single part
+pub(crate) struct PartData {
+    // 1-indexed 
+    pub(crate) part_number: u64,
+    pub(crate) data: Bytes,
+}
+
+/// The `ReadPart` trait allows for reading data from an `InputStream` and packaging the raw
+/// data into `PartData` which carries additional metadata needed for uploading a part.
+pub(crate) trait ReadPart {
+    /// Request the next "part" of data.
+    ///
+    /// When there is no more data readers should return `Ok(None)`.
+    /// NOTE: Implementations are allowed to return data in any order and consumers are
+    /// expected to order data by the part number.
+    fn next_part(
```

Comment:
Just need to warn, this API, where the uploader "pulls/reads" data from the user, worked great in aws-c-s3 until we hit the use case where a user starts tons of uploads, but doesn't actually control when the data will arrive for any particular file. We ended up adding an alternate "push/write" API to cover this use case. See the Issue description in this PR: https://github.com/awslabs/aws-c-s3/pull/418

An API that covered both cases would probably work something like:
- 1 async call where user requests permission to submit another part. The async call completes when the uploader is ready to accept more data (maybe it returns a buffer from a pool, for them to fill in?).
- another call to submit the filled-in buffer

Reply:
Yeah we talked about this the other day briefly where you mentioned this use case. I think we need to flesh that out more and consider this design incomplete until we cover it.

Right now this is basically the view/abstraction that a single uploader task will see (be more clear in a future PR but you can see in-branch [here](https://github.com/smithy-lang/smithy-rs/blob/atodd/s3-mpu/aws/hll/aws-s3-transfer-manager/src/upload.rs#L232-L237) and [here](https://github.com/smithy-lang/smithy-rs/blob/atodd/s3-mpu/aws/hll/aws-s3-transfer-manager/src/upload.rs#L320-L325)). This is obviously simplistic as it doesn't cover users needing/wanting to control checksums nor the ability for users to control which uploads have data ready.